### PR TITLE
[part of #832] use db_id for iam_projects and iam_[staged_]project_rules

### PR DESCRIPTION
**Pull request by srenatus**
### :nut_and_bolt: Description

Any (staged, applied) rules should now only reference `iam_projects` via its `db_id` column.

The primary key switch (`id` -> `db_id`) for `iam_projects` is to be done after the sibling tasks of this one.

### :chains: Related Resources

#832

### :white_check_mark: Checklist

- [X] Necessary tests added/updated?
- [ ] Necessary docs added/updated?
- [X] Code actually executed?
- [X] Vetting performed (unit tests, lint, etc.)?

### CODE REVIEW

File: components/authz-service/storage/postgres/migration/sql/54_use_project_db_id_in_references_of_project_rules.up.sql [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+ALTER TABLE iam_project_rules
+    DROP CONSTRAINT iam_project_rules_project_id_fkey;
+ALTER TABLE iam_staged_project_rules
+    DROP CONSTRAINT iam_staged_project_rules_project_id_fkey;
+
+-- When we're done fixing all references, we'll change the
+-- primary key of iam_projects; now, we only add a uniqueness
+-- constraint on what's going to be the new PK: db_id.
+ALTER TABLE iam_projects
+    ADD UNIQUE (db_id);
```
Comment (srenatus) [inline components/authz-service/storage/postgres/migration/sql/54_use_project_db_id_in_references_of_project_rules.up.sql]: This allows us to update the references piece-by-piece. When we're done with the other tables, we can finally
```
ALTER TABLE iam_projects
    DROP CONSTRAINT iam_projects_pkey;
ALTER TABLE iam_projects
    ADD PRIMARY KEY (db_id);
ALTER TABLE iam_projects
    ADD UNIQUE (id);
```

File: components/authz-service/storage/postgres/migration/sql/54_use_project_db_id_in_references_of_project_rules.up.sql [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+
+-- fix functions
+CREATE OR REPLACE FUNCTION
+  query_rules_for_project(_project_id TEXT, _project_filter TEXT[])
+  RETURNS setof json AS $$
+
+  -- fetch staged rules
+   SELECT json_build_object(
+    'id', r.id,
+    'project_id', p.id,
```
Comment (srenatus) [inline components/authz-service/storage/postgres/migration/sql/54_use_project_db_id_in_references_of_project_rules.up.sql]: ℹ️ What this returns is the same, but it now needs to join the other table to get the text ID of the project.

File: components/authz-service/storage/postgres/migration/sql/54_use_project_db_id_in_references_of_project_rules.up.sql [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+        'value', rc.value,
+        'operator', rc.operator,
+        'attribute', rc.attribute
+      )
+    )
+  ) AS rule
+  FROM iam_staged_project_rules AS r
+  INNER JOIN iam_staged_rule_conditions AS rc 
+  ON rc.rule_db_id=r.db_id
+  INNER JOIN iam_projects AS p
```
Comment (srenatus) [inline components/authz-service/storage/postgres/migration/sql/54_use_project_db_id_in_references_of_project_rules.up.sql]: ℹ️ Here's the added join, the same stuff gets repeated over and over again below...

File: components/authz-service/storage/postgres/migration/sql/54_use_project_db_id_in_references_of_project_rules.up.sql [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+$$ LANGUAGE sql;
+
+-- NEW: it takes a project's db_id as first argument
+-- This way, we can keep some of the other functions unchanged.
+CREATE OR REPLACE FUNCTION
+  projects_match_for_rule(_project_db_id INTEGER, _projects_filter TEXT[])
+  RETURNS BOOLEAN AS $$
+    SELECT array_length(_projects_filter, 1) IS NULL
+    OR EXISTS (SELECT 1 FROM iam_projects WHERE db_id=_project_db_id AND id=ANY(_projects_filter))
+$$ LANGUAGE SQL;
```
Comment (srenatus) [inline components/authz-service/storage/postgres/migration/sql/54_use_project_db_id_in_references_of_project_rules.up.sql]: ℹ️ In a few places, we have `WHERE projects_match_for_rule(project_id, _project_filter)` -- this new function, with a different type signature, works as a drop-in replacement, how that the `project_id` column of (staged) rules is INTEGER.

File: components/authz-service/storage/postgres/migration/sql/54_use_project_db_id_in_references_of_project_rules.up.sql [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+
+CREATE OR REPLACE FUNCTION update_rule(_in_rule_id text, _in_project_id text, _in_name text, _in_type text, projects_filter TEXT[])
```
Comment (srenatus) [inline components/authz-service/storage/postgres/migration/sql/54_use_project_db_id_in_references_of_project_rules.up.sql]: ℹ️ Only mildly adjusted. Besides sprinkling a few `project_db_id(_in_project_id)` over this, I've replaced
```
EXISTS (SELECT x, y FROM ...)
```
with
```
EXISTS (SELECT 1 FROM ...)
```
since I believe the fields we select don't matter at all there.

File: components/authz-service/storage/postgres/migration/sql/README.md [hunk 6]
```diff
@@ -53,3 +53,4 @@
 - [`51_change_project_rules_query.up.sql`](51_change_project_rules_query.up.sql)
 - [`52_add_db_id_members.up.sql`](52_add_db_id_members.up.sql)
 - [`53_add_db_id_to_iam_policies.up.sql`](53_add_db_id_to_iam_policies.up.sql)
+- [`54_use_project_db_id_in_references_of_project_rules.up.sql`](54_use_project_db_id_in_references_of_project_rules.up.sql)
```
Comment (srenatus) [inline components/authz-service/storage/postgres/migration/sql/README.md]: ⚠️ Will clash with #855, let's see who merges first 🐌

File: components/authz-service/storage/v2/postgres/postgres.go [hunk 7]
```diff
@@ -1066,7 +1068,7 @@ func (p *pg) DeleteRule(ctx context.Context, id string) error {
 		res, err := tx.ExecContext(ctx,
```
Comment (srenatus) [inline components/authz-service/storage/v2/postgres/postgres.go]: ℹ️ Thanks to the new helper function `projects_match_for_rule(INTEGER, TEXT[])`, the changes to `DeleteRule` at only drive-by nitpicks.

File: components/authz-service/storage/postgres/migration/sql/54_use_project_db_id_in_references_of_project_rules.up.sql [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+  WHERE projects_match_for_rule(p.id, _project_filter)
+  GROUP BY r.id, p.id, r.name, r.type;
+
+$$ LANGUAGE sql;
+
+CREATE OR REPLACE FUNCTION
+  query_staged_or_applied_rule(_rule_id TEXT, _project_filter TEXT[])
+  RETURNS json AS $$
+
+  -- check for applied rule
```
Comment (msorens) [inline components/authz-service/storage/postgres/migration/sql/54_use_project_db_id_in_references_of_project_rules.up.sql]: For posterity, this comment should read: _check for staged rule_
